content/blog/r-mpi-part-one.md

```diff
@@ -0,0 +1,220 @@
+---
+author: "Wes Hinsley"
+date: 2023-06-22
+title: MPI from R - Part One
+best: false
+tags:
+ - R
+ - MPI
+ - HPC
+---
+
+[MPI](https://en.wikipedia.org/wiki/Message_Passing_Interface) is a 
+standard for allowing a team of processes to solve a problem together. Various
+implementations of that standard exist; [MPICH](https://www.mpich.org/) is
+perhaps the most well known. [Microsoft MPI](https://github.com/Microsoft/Microsoft-MPI)
+mimics MPICH on Windows, as does [Intel MPI](https://www.intel.com/content/www/us/en/developer/tools/oneapi/mpi-library.html#gs.1crzpx)
+on various platforms. Typically you'd then include these libraries in projects 
+written in C, C++, or Fortran.
+
+Furthermore, recent releases of [Rtools](https://cran.r-project.org/bin/windows/Rtools/)
+have included support for Microsoft MPI. Here, we're going to use that to write some
+MPI code in R for Windows, and run it with multiple local processes, and then using
+multiple nodes on our departmental MS-HPC cluster.
+
+# First steps.
+
+First we need an MPI library. On Windows, installing
+the latest [Rtools](https://cran.r-project.org/bin/windows/Rtools/) and
+[MS-MPI](https://learn.microsoft.com/en-us/message-passing-interface/microsoft-mpi)
+will do it. Test by running `mpiexec` in a terminal, and on Windows check that it really is
+Microsoft's `mpiexec`. Intel's has the same name, and comes with the Intel C++ compiler, but
+Microsoft's must come first in your path, as we are need the `mpiexec` executable we use
+to match the library in Rtools.
+
+On linux, `sudo apt-get install mpich.` is enough; on Mac, [download MPICH](http://www.mpich.org/downloads/) 
+and it will install with homebrew. 
+
+# What defines an MPI program
+
+An MPI program involves a number of processes, all running the same single executable code.
+The processes would traditionally have been on different computers - one process on each - 
+but they can also be stacked on the same compute node, or spread with some number of 
+processes across some number of nodes. 
+
+The program must have exactly one call to `MPI_Init`, during which all the processes handshake, 
+and by consensus they decide their ids - integers starting at zero - for each process in the 
+family. The id is known as the _rank_, and the number of processes is the _size_. After that
+initialise step, each process knows its own rank, and can use it to decide a 
+subset of the total work to be done for example.
+
+Finally, exactly one `MPI_Finalize` call needs to happen (on every process) when all the MPI 
+work is finished, if you want a neat and successful exit.
+
+# Parallel Hello World
+
+Here come the bits of C++ code, which I'm writing in an R package called `mpitest`.
```

Comment:
Is this the typical way to use MPI with R? You wouldn't use it to parallelise R code itself?

Reply:
Yeah - MPI has to come in that way, but in the end we should be able to call those things either in more C++ core, or from R. I'll say something on that...